rpc/lib/client/ws_client.go

```diff
@@ -1,160 +1,390 @@
-	*websocket.Conn
-	ResultsCh chan json.RawMessage // closes upon WSClient.Stop()
-	ErrorsCh  chan error           // closes upon WSClient.Stop()
+
+	PingPongLatencyTimer metrics.Timer
+	sentLastPingAt       time.Time
+
+	// user facing channels, closed only when the client is being stopped.
+	ResultsCh chan json.RawMessage
+	ErrorsCh  chan error
+
+	// internal channels
+	send               chan types.RPCRequest // user requests
+	backlog            chan types.RPCRequest // stores a single user request received during a conn failure
+	reconnectAfter     chan error            // reconnect requests
+	receiveRoutineQuit chan struct{}         // a way for receiveRoutine to close writeRoutine
+
+	reconnecting bool
+
+	wg  sync.WaitGroup
+	mtx sync.RWMutex
+
+	// Time allowed to read the next pong message from the server.
+	pongWait time.Duration
+
+	// Send pings to server with this period. Must be less than pongWait.
+	pingPeriod time.Duration
 }
 
-// create a new connection
-func NewWSClient(remoteAddr, endpoint string) *WSClient {
+// NewWSClient returns a new client.
+func NewWSClient(remoteAddr, endpoint string, options ...func(*WSClient)) *WSClient {
 	addr, dialer := makeHTTPDialer(remoteAddr)
-	wsClient := &WSClient{
-		Address:  addr,
-		Dialer:   dialer,
-		Endpoint: endpoint,
-		Conn:     nil,
+	c := &WSClient{
+		Address:              addr,
+		Dialer:               dialer,
+		Endpoint:             endpoint,
+		PingPongLatencyTimer: metrics.NewTimer(),
+		pongWait:             defaultPongWait,
+		pingPeriod:           defaultPingPeriod,
+	}
+	c.BaseService = *cmn.NewBaseService(nil, "WSClient", c)
+	for _, option := range options {
+		option(c)
+	}
+	return c
+}
+
+// PingPong allows changing ping period and pong wait time. If ping period
+// greater or equal to pong wait time, panic will be thrown.
+func PingPong(pingPeriod, pongWait time.Duration) func(*WSClient) {
+	return func(c *WSClient) {
+		if pingPeriod >= pongWait {
+			panic(fmt.Sprintf("ping period (%v) must be less than pong wait time (%v)", pingPeriod, pongWait))
+		}
+		c.pingPeriod = pingPeriod
+		c.pongWait = pongWait
 	}
-	wsClient.BaseService = *cmn.NewBaseService(nil, "WSClient", wsClient)
-	return wsClient
 }
 
-func (wsc *WSClient) String() string {
-	return wsc.Address + ", " + wsc.Endpoint
+// String returns WS client full address.
+func (c *WSClient) String() string {
+	return fmt.Sprintf("%s (%s)", c.Address, c.Endpoint)
 }
 
-// OnStart implements cmn.BaseService interface
-func (wsc *WSClient) OnStart() error {
-	wsc.BaseService.OnStart()
-	err := wsc.dial()
+// OnStart implements cmn.Service by dialing a server and creating read and
+// write routines.
+func (c *WSClient) OnStart() error {
+	err := c.dial()
 	if err != nil {
 		return err
 	}
-	wsc.ResultsCh = make(chan json.RawMessage, wsResultsChannelCapacity)
-	wsc.ErrorsCh = make(chan error, wsErrorsChannelCapacity)
-	go wsc.receiveEventsRoutine()
+
+	c.ResultsCh = make(chan json.RawMessage)
```

Comment:
it's still hard for me to choose between channels and callbacks. should we use callbacks while delivering results? maybe. google uses callback style in his pubsub https://godoc.org/cloud.google.com/go/pubsub#hdr-Receiving